agents/tools/upstream_search.py

```diff
@@ -0,0 +1,140 @@
+                            result=UpstreamSearchResult.NOT_POSSIBLE,
+                            repository_url=None,
+                            related_commits=None
+                        ))
+                    repos = await response.json()
+
+                # until we have solid reference to upstream repository through for example VCS
+                # spec file tag, this is the best we can do
+                post_params = {"url": repos[0], "text": tool_input.description}
+                if tool_input.date is not None:
+                    post_params["date"] = tool_input.date
+                async with session.post(f"{os.environ['UPSTREAM_SEARCH_API_URL']}/find_commit",
+                                        json=post_params, timeout=240) as response:
+                    if response.status != 200:
+                        logger.debug("Searching did not yield commits. status %d response %s",
+                                      response.status, await response.text())
+                        return UpstreamSearchToolOutput(UpstreamSearchToolResult(
+                            result=UpstreamSearchResult.NOT_FOUND,
+                            repository_url=None,
+                            related_commits=None
+                        ))
+                    commits = await response.json()
+        except asyncio.TimeoutError:
+            raise ToolError("Timeout occured while contacting upstream-search backend")
+        except Exception as e:
+            raise ToolError(f"Unexpected internal error occured while contacting backend {e}")
+
+        def format_commits(repos: list[str], commits: list[str]) -> list[str]:
+            # the agent struggles with finding the right format to retrieve the
+            # patch from git repository hosting, so we try to resolve this for him if possible
+            right_format_prefix = ""
+            right_format_suffix = ""
+            try:
+                parsed_url = urlparse(repos[0])
+                if not parsed_url.path.endswith(".git"):
+                    return
+                if parsed_url.hostname == "github.com":
+                    parsed_url = parsed_url._replace(
+                        path=parsed_url.path.replace(".git", "/commit/"))
+                    right_format_prefix = parsed_url.geturl()
+                    right_format_suffix = ".patch"
+                elif parsed_url.hostname.startswith("gitlab"):
+                    parsed_url = parsed_url._replace(
+                        path=parsed_url.path.replace(".git", "/-/commit/"))
+                    right_format_prefix = parsed_url.geturl()
+                    right_format_suffix = ".patch"
+            except ValueError:
+                # can happen and we will fallback to letting agent guess the right format
+                return commits
+
+            return [right_format_prefix + commit + right_format_suffix for commit in commits]
+
+        commits = format_commits(repos, commits)
```

Comment:
This is not what I meant, rather something like:

```suggestion
        def get_patch_url(commit):
            parsed_url = urlparse(repos[0])
            if not parsed_url.path.endswith(".git"):
                return commit
            prefix = "/-" if parsed_url.hostname.startswith("gitlab") else ""
            path = f"{prefix}/commit/{commit}.patch"
            return parsed_url._replace(path=parsed_url.path.replace(".git", path)).geturl()

        commits = [get_patch_url(commit) for commit in commits]
```

I'm not sure what in the original code can raise `ValueError`, I don't see anything.

Reply:
`ValueError` handling is here out of paranoia. urlparse can throw it if url is not properly formatted. Your suggestion is certainly shorter and more elegant, but we can not unfortunately let prefix be `"/-" if parsed_url.hostname.startswith("gitlab") else ""` because not all upstreams are gitlab or github.

See this for example `http://thekelleys.org.uk/git/dnsmasq.git`. The proper url for commit there is `https://thekelleys.org.uk/gitweb/?p=dnsmasq.git;a=patch;h=<commit_id>` and this assignment would give agent `http://thekelleys.org.uk/commit/<commit_id>.patch` which does not work for gitweb and there can be more non standard urls where we have better chance if we give the agent plain hash and let it guess the proper url path.